pyomo/repn/plugins/gams_writer.py

```diff
@@ -0,0 +1,512 @@
+        put_results = io_options.pop("put_results", None)
+
+        if len(io_options):
+            raise ValueError(
+                "ProblemWriter_gams passed unrecognized io_options:\n\t" +
+                "\n\t".join("%s = %s"
+                            % (k,v) for k,v in io_options.iteritems()))
+
+        if solver is not None:
+            if solver.upper() not in valid_solvers:
+                raise ValueError("ProblemWriter_gams passed unrecognized "
+                                 "solver: %s" % solver)
+
+        if mtype is not None:
+            valid_mtypes = set([
+                'lp', 'qcp', 'nlp', 'dnlp', 'rmip', 'mip', 'rmiqcp', 'rminlp',
+                'miqcp', 'minlp', 'rmpec', 'mpec', 'mcp', 'cns', 'emp'])
+            if mtype.lower() not in valid_mtypes:
+                raise ValueError("ProblemWriter_gams passed unrecognized "
+                                 "model type: %s" % mtype)
+            if (solver is not None and
+                mtype.upper() not in valid_solvers[solver.upper()]):
+                raise ValueError("ProblemWriter_gams passed solver (%s) "
+                                 "unsuitable for given model type (%s)"
+                                 % (solver, mtype))
+
+        if output_filename is None:
+            output_filename = model.name + ".gms"
+
+        if symbolic_solver_labels and (labeler is not None):
+            raise ValueError("ProblemWriter_gams: Using both the "
+                             "'symbolic_solver_labels' and 'labeler' "
+                             "I/O options is forbidden")
+
+        if symbolic_solver_labels:
+            var_labeler = con_labeler = AlphaNumericTextLabeler()
+        elif labeler is None:
+            var_labeler = NumericLabeler('x')
+            con_labeler = NumericLabeler('c')
+        else:
+            var_labeler = con_labeler = labeler
+
+        var_list = []
+        symbolMap = SymbolMap()
+
+        def var_recorder(obj):
+            ans = var_labeler(obj)
+            var_list.append(ans)
+            return ans
+
+        def var_label(obj):
+            if obj.is_fixed():
+                return str(value(obj))
+            return symbolMap.getSymbol(obj, var_recorder)
+
+        # when sorting, there are a non-trivial number of
+        # temporary objects created. these all yield
+        # non-circular references, so disable GC - the
+        # overhead is non-trivial, and because references
+        # are non-circular, everything will be collected
+        # immediately anyway.
+        with PauseGC() as pgc:
+            try:
+                if isinstance(output_filename, string_types):
+                    output_file = open(output_filename, "w")
+                else:
+                    # Support passing of stream such as a StringIO
+                    # on which to write the model file
+                    output_file = output_filename
+                if isinstance(model, IBlockStorage):
```

Comment:
I'm wondering if I'm understanding this problem correctly: is the issue that if there are multiple processes running at the same time, if they both use this ComponentData.labeler list, then they will confuse each other by trampling on the other's storage location for their labeler? And thus there needs to be a way to pass a labeler that is unique to each process?

If so, I see two options. One is another sort of workaround, where ComponentData.labeler is a dictionary instead of a list, and the labeler is added to this dictionary with a key that is unique to each process. Of course then you have to decide how to generate keys, and me not knowing many options here I can think of making a big random number.

The other option I see, as you said, is to pass the labeler as a keyword to to_string. This would require updating all the different definitions of to_string, for which there are many when it comes to all the different types of expressions. They all need to be updated so that the labeler keyword can be passed along until it reaches a Var, who will act accordingly if it is passed a labeler. Is this the option you prefer?